controller-server/src/test/java/com/yahoo/vespa/hosted/controller/maintenance/MetricsReporterTest.java

```diff
@@ -494,6 +497,60 @@ public void tenant_counter() {
         assertEquals(1, metrics.getMetric(d -> "trial".equals(d.get("plan")), MetricsReporter.TENANT_METRIC).get());
     }
 
+    @Test
+    public void upgrade_duration_inside_block_window() {
+        ApplicationPackage pkg = new ApplicationPackageBuilder().region("us-west-1")
+                                                                // window 1
+                                                                .blockChange(false, true, "mon-tue", "2-9", "CET")
+                                                                // window 2
+                                                                .blockChange(false, true, "mon-tue", "1-8,10-11", "CET")
+                                                                // window 3
+                                                                .blockChange(false, true, "wed-thu", "0-23", "CET")
+                                                                // window 4 (does not apply to upgrade)
+                                                                .blockChange(true, false, "mon-sun", "0-7", "CET")
+                                                                .build();
+
+        Instant mondayNight = Instant.parse("2021-12-13T23:00:00.00Z");
+        DeploymentTester tester = new DeploymentTester().at(mondayNight);
+        MetricsReporter reporter = createReporter(tester.controller());
+        DeploymentContext context = tester.newDeploymentContext();
+        Supplier<Duration> metric = () -> {
+            reporter.maintain();
+            return Duration.ofSeconds(metrics.getMetric(context.instanceId(),MetricsReporter.DEPLOYMENT_OVERDUE_UPGRADE)
+                                             .get().longValue());
+        };
+
+        // Deploy completely once
+        context.submit(pkg).completeRollout();
+
+        // System is upgraded, triggering upgrade of application
+        tester.controllerTester().upgradeSystem(Version.fromString("7.0"));
+        tester.upgrader().maintain();
+
+        // Start production job for upgrade, without completing it
+        context.runJob(systemTest)
+               .runJob(stagingTest)
+               .triggerJobs()
+               .assertRunning(productionUsWest1);
+        assertEquals("Upgrade is not overdue yet", Duration.ZERO, metric.get());
+
+        // Upgrade continues into block window
+        tester.clock().advance(Duration.ofHours(3)); // Tuesday at 02:00 (03:00 CET)
+        assertEquals("Upgrade is overdue measured relative to window 2", Duration.ofHours(2), metric.get());
+
+        tester.clock().advance(Duration.ofHours(6)); // Tuesday at 08:00 (09:00 CET)
```

Comment:
I find it more natural to take the union of all block windows, rather than to split them. Is there a particular reason for this choice?